### DISCUSSION TITLE
What is the difference between 'Visit' method and 'DefaultVisit' method in SyntaxVisitors

### DISCUSSION THREAD

**Discussion by sanmuru** [dotnet/roslyn#60108] [Q&A]
I am trying to declare my own SyntaxVisitor derived from [`CSharpSyntaxVisitor`](/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxVisitor.cs), and I wondered which method(**Visit** or **DefaultVisit**) I should choose to override. As there is a lack of documentation for Roslyn, I try to search them through whole solution and find out nearly all of **DefaultVisit** references are from `CSharpSyntaxVisitor` itself, which means **DefaultVisit** is used as `protected` method.
But why is it called **DefaultVisit**? And what is the difference between these two methods?
```CSharp
public abstract partial class CSharpSyntaxVisitor<TResult>
{
    public virtual TResult? Visit(SyntaxNode? node)
    {
        if (node != null)
        {
            return ((CSharpSyntaxNode)node).Accept(this);
        }

        // should not come here too often so we will put this at the end of the method.
        return default;
    }

    public virtual TResult? DefaultVisit(SyntaxNode node)
    {
        return default;
    }
}

public abstract partial class CSharpSyntaxVisitor
{
    public virtual void Visit(SyntaxNode? node)
    {
        if (node != null)
        {
            ((CSharpSyntaxNode)node).Accept(this);
        }
    }

    public virtual void DefaultVisit(SyntaxNode node)
    {
    }
}
```

**Comment by CyrusNajmabadi** [comment]
DefaultVisit is called for every node whose specific Visit method you don't override. It's a way of allowing you to specify a fallback behavior you apply to everything if they don't match the specific behaviors you apply to a subset of node types. 

For example, you might think you only need to handle 5 node types in your visitor. So you override those 5 methods. But you also override DefaultVisit to throw if hit. That way if you missed something, you find out with a crash that can include all the information you need.
